Added toString implementation to DelegatingGenericResponse

bbaugher opened the pull request:
What was changed? Why is this necessary?
----------------------------------------

Dumb but nice to have, added toString implementation to help testing and debugging. 

I originally did this while playing around with the client and implemented it on what the client sees (DelegatingGenericResponse) but I'm thinking this should instead be implemented on the interface (GenericResponse) instead otherwise you would likely want to implement toString on BuiltGenericResponse to be consistent. Let me know if you have a preference.

I also effectively implemented toString for raw Response otherwise you just get a reference to ApacheHttpClient4Dot3Engine on the client side and from what I can tell BuiltResponse on server side. I can't decide if I should try to provide a pull to resteasy to implement toString for those classes or just leave it here. Again let me know if you have a preference.

How was it tested?
----

Built locally and ran with a client implementation

How to test
----

*This is bare minimum acceptable testing*

- [ ] `mvn clean install -U`

Reviewers
----

- [x] [John Leacox](https://github.com/johnlcox)
- [x] [Sundeep Paruvu](https://github.com/sparuvu)
- [x] [Nimesh Subramanian](https://github.com/nimeshsubramanian)
- [ ] [Supriya Lal](https://github.com/lal-s)
- [ ] [Brian van de Boogaard](https://github.com/b-boogaard)

sparuvu commented on jaxrs-genericresponse/src/main/java/com/cerner/beadledom/jaxrs/DelegatingGenericResponse.java:
```diff
@@ -25,6 +25,7 @@
   private final T body;
   private final Class<?> bodyClass;
   private final ErrorBody errorBody;
+
```
can you please remove this unnecessary extra empty line.

bbaugher replied:
Done, https://github.com/cerner/beadledom/pull/28/commits/08ddce40ddfc37c67b994dbde0f54f91d014f62b

sparuvu commented:
can you please update the changelog with this change ?

sparuvu commented:
`JAXRS-genericResponse` is in Java 6 so we can't have the default methods on the interface :(. I think, Implementing`toString` on the `DelegatingGenericResponse` is fine I guess. It would be great if we could add `toString` on `BuiltGenericResponse` as well.

bbaugher commented:
Added to changelog, https://github.com/cerner/beadledom/pull/28/commits/115608019cabb55acb48685d31eceda510035c37

Implemented toString in BuiltGenericResponse, https://github.com/cerner/beadledom/pull/28/commits/159a76e1d482a8ebe12cc253aedbaccc74880ebf

nimeshsubramanian commented:
+1

sparuvu commented:
I am wondering if we should add a test for each class 🤔

b-boogaard commented:
+1

johnlcox commented on resteasy-genericresponse/src/main/java/com/cerner/beadledom/resteasy/BuiltGenericResponse.java:
```diff
@@ -255,4 +256,15 @@ public Link getLink(String relation) {
+  @Override
+  public String toString() {
+    return "BuiltGenericResponse{body=" + body + ", bodyClass=" + bodyClass + ", errorBody=" + errorBody
+            + ", rawResponse=Response{status=" + rawResponse.getStatus() + ", mediaType="
```
Rather than duplicating the `rawResponse` String code, it might be better to have a class like `ResponseToStringWrapper` that only has a `toString` method on it.

bbaugher replied:
Done, https://github.com/cerner/beadledom/pull/28/commits/3a04e401c758fe3f8e1b39405bf0581995634f41

bbaugher commented:
@sparuvu If you would like, because theres almost no logic I had left it out

johnlcox commented on jaxrs-genericresponse/src/main/java/com/cerner/beadledom/jaxrs/ResponseToStringWrapper.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Simple class to implement {@link Response} to String.
+ */
+public class ResponseToStringWrapper {
```
This is more of a utility class than a wrapper. I was imagining a class that takes in `Response` to it's constructor and then implements `Object#toString()`.

sparuvu commented:
> @sparuvu If you would like, because theres almost no logic I had left it out

I prefer to have this code covered.

bbaugher commented:
Added tests and converted the wrapper to implement toString and take in Response in the constructor, let me know if theres anything else

johnlcox commented on resteasy-genericresponse/src/test/scala/com/cerner/beadledom/resteasy/BuiltGenericResponseSpec.scala:
```diff
@@ -436,5 +440,30 @@ class BuiltGenericResponseSpec
+        val rawResponse = mock[Response]
+        when(rawResponse.getStatus).thenReturn(200)
+        when(rawResponse.getMediaType).thenReturn(MediaType.APPLICATION_JSON_TYPE)
+        when(rawResponse.getDate).thenReturn(new Date(0L))
```
Looks like the dates are failing on Travis due to timezone differences. Might be best to just get the expected dates programmatically as well.

bbaugher replied:
Should be fixed now, https://github.com/cerner/beadledom/pull/28/commits/bd07cae96b21c4610e955327b3f221aebbe23bd8

johnlcox commented:
+1
